### ISSUE TITLE
EVP_SKEY

### ISSUE THREAD

**Issue by slontis** [openssl/openssl#19754]
Currently we pass byte arrays around when dealing with symmetric keys.
There should be an opaque EVP_SKEY object that represents this object instead...
The actual representation of the object should be provider specific.
In most cases it would just be a byte array, but it could also just be a handle/id  (for pkcs11).

````
EVP_SKEY_CTX *EVP_SKEY_CTX_new(void);
void EVP_SKEY_CTX_free(EVP_SKEY_CTX *sctx);
int EVP_SKEY_keygen_init(EVP_SKEY_CTX *sctx, int size, const OSSL_PARAM params[]);
int EVP_SKEY_keygen(EVP_SKEY_CTX *sctx, EVP_SKEY **ppkey);
int EVP_SKEY_CipherInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                        EVP_SKEY *dkey, const unsigned char *iv,
                        int enc, const OSSL_PARAM params[]);
````

I imagine it would require new dispatch table methods for encrypt and decrypt.

**Comment by slontis** [comment]
@romen I know this was a long time ago, but I believe you looked into this at some stage?

**Comment by beldmit** [comment]
@simo5 you should like this approach

**Comment by t8m** [comment]
> I imagine it would require new dispatch table methods for encrypt and decrypt.

Not just this, but also various KDFs, KEMs, MACs....

**Comment by simo5** [comment]
Yeah absolutely agree, in fact I already felt the need for this in the pkcs11 provider because KDFs or KEMs generate keys.

Not all tokens allow for exporting them as bytes (or at all), but can definitely use them in later operations if they are represented as a provider key because those keys can be used by the token for symmetric operations.

Current code just return an error for operations that require returning a key as bytes but the token does not allow it ... which is less than ideal.

**Comment by paulidale** [comment]
> I imagine it would require new dispatch table methods for encrypt and decrypt.

It might not....

**Comment by slontis** [comment]
There is no way to pass objects via OSSL_PARAM, so I dont see how that would work.

**Comment by slontis** [comment]
For setting peers keys etc, we pass them as pointers in the methods..

**Comment by levitte** [comment]
> ``` C
> int EVP_SKEY_CipherInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
>                         EVP_SKEY *dkey, const unsigned char *iv,
>                         int enc, const OSSL_PARAM params[]);
> ```

May I suggest an operation based API like we have for MAC, KDF, etc?

``` C
int EVP_CIPHER_init(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                    EVP_SKEY *dkey, const unsigned char *iv,
                    int enc, const OSSL_PARAM params[]);
int EVP_CIPHER_update(EVP_CIPHER_CTX *ctx, const unsigned char *data, size_t datalen);
int EVP_CIPHER_final(EVP_CIPHER_CTX *ctx,
                     unsigned char *out, size_t *outl, size_t outsize);
```

We certainly do have the API space for these, and that would increase consistency with later development...

**Comment by levitte** [comment]
> There is no way to pass objects via OSSL_PARAM, so I dont see how that would work.

Correct.  OSSL_PARAM is meant for extra optional (across all implementations) parameters.

**Comment by levitte** [comment]
> > I imagine it would require new dispatch table methods for encrypt and decrypt.
> 
> It might not....

It might...

``` C
OSSL_CORE_MAKE_FUNC(int, cipher_encrypt_init, (void *cctx,
                                                  const unsigned char *key,
                                                  size_t keylen,
                                                  const unsigned char *iv,
                                                  size_t ivlen,
                                                  const OSSL_PARAM params[]))
```

It might be quite confusing if `key` was suddenly expected to be a pointer to a structure...

**Comment by paulidale** [comment]
`key` would be a pointer to a provider created blob of data.  What's in the blob isn't specified.

This would require some rework and more thought than I've currently put in…

**Comment by levitte** [comment]
> `key` would be a pointer to a provider created blob of data. What's in the blob isn't specified.

Although, I think that everyone expects it to be a bytestring. But you may be right that it's not as difficult a transition as I feared. 
However... 

> This would require some rework and more thought than I've currently put in…

Yeah, my first thought is how this would interact with KDFs...

**Comment by mattcaswell** [comment]
If we are having new APIs for this, then we should consider how AEAD fits into it (the currently existing way of doing AEAD is fairly horrible).

**Comment by paulidale** [comment]
AEAD should be a completely different operation type.
We missed that boat.

**Comment by beldmit** [comment]
I think this can be closed now as we have a basic implementation already.
